src/tsdb.c

```diff
@@ -183,10 +187,94 @@ void CleanLastDeletedSeries(RedisModuleCtx *ctx, RedisModuleString *key) {
                 RedisModule_CloseKey(seriesKey);
             }
         }
+
+        RedisModule_FreeThreadSafeContext(ctx);
     }
     freeLastDeletedSeries();
 }
 
+void RenameSeriesFrom(RedisModuleCtx *ctx, RedisModuleString *key) {
+    // keep in global variable for RenameSeriesTo() and increase recount
+    RedisModule_RetainString(NULL, key);
+    renameFromKey = key;
+}
+
+void RenameSeriesTo(RedisModuleCtx *ctx, RedisModuleString *keyTo) {
+    // Try to open the series
+    Series *series;
+    RedisModuleKey *key = NULL;
+    const int status = SilentGetSeries(ctx, keyTo, &key, &series, REDISMODULE_READ);
+    if (!status) { // Not a timeseries key
+        goto cleanup;
+    }
+
+    // Reindex key by the new name
+    RemoveIndexedMetric(ctx, renameFromKey, series->labels, series->labelsCount);
+    IndexMetric(ctx, keyTo, series->labels, series->labelsCount);
+
+    // A destination key was renamed
+    if (series->srcKey) {
+        Series *srcSeries;
+        RedisModuleKey *srcKey;
+        const int status =
+            SilentGetSeries(ctx, series->srcKey, &srcKey, &srcSeries, REDISMODULE_WRITE);
+        if (!status) {
+            const char *srcKeyName = RedisModule_StringPtrLen(series->srcKey, NULL);
+            RedisModule_Log(
+                ctx, "warning", "couldn't open key or key is not a Timeseries. key=%s", srcKeyName);
+            goto cleanup;
+        }
+
+        // Find the rule in the source key and rename the its destKey
+        CompactionRule *rule = srcSeries->rules;
+        while (rule) {
+            const char *xxx = RedisModule_StringPtrLen(renameFromKey, NULL);
+            const char *yyy = RedisModule_StringPtrLen(rule->destKey, NULL);
+
+            if (RedisModule_StringCompare(renameFromKey, rule->destKey) == 0) {
+                RedisModule_FreeString(NULL, rule->destKey);
+                RedisModule_RetainString(NULL, keyTo);
+                rule->destKey = keyTo;
+                break; // Only one src can point back to destKey
+            }
+            rule = rule->nextRule;
+        }
+        RedisModule_CloseKey(srcKey);
+    }
+
+    // A source key was renamed need to rename the srcKey on all the destKeys
+    if (series->rules) {
+        CompactionRule *rule = series->rules;
+        Series *destSeries;
+        RedisModuleKey *destKey;
+        while (rule) {
+            const int status =
+                SilentGetSeries(ctx, rule->destKey, &destKey, &destSeries, REDISMODULE_WRITE);
+            if (!status) {
+                const char *destKeyName = RedisModule_StringPtrLen(rule->destKey, NULL);
+                RedisModule_Log(ctx,
+                                "warning",
+                                "couldn't open key or key is not a Timeseries. key=%s",
+                                destKeyName);
```

Comment:
no need for error handling? goto cleanup?

Reply:
@MeirShpilraien This can happen if we have a rule with a missing downsampled key, in this case we want to skip handling it and move to the next rule

Reply:
@danni-m so you need at least a continue here ...